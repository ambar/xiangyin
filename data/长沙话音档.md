# 长沙话音档

## CUHK 漢語多功能字庫的错误汇总

1. #1 n̍ 读成 li
1. #2 i 发间不太准，重复，用 #990 更好
1. #48 uai 读成 5，与 #49 一样
1. #154 piẽ 读成 5
1. #161 pan 读成 4
1. #183 pʰa 调为陰平33，错标为上聲41
1. #199 pʰau 读成 4
1. #202 pʰiau 读成 1
1. #212 pʰan 读成 ang
1. #222 pʰoŋ 读成 1
1. #236 mo 读成 2
1. #249 mei 读成 4
1. #263 məu 读成了 mau
1. #268 miẽ 读成 5
1. #277 mən 读成 2
1. #282 moŋ 读成 men1
1. #291 fu 读成 4
1. #362 təu 读成 1
1. #401 tʰa 读成 4
1. #426 tʰiẽ 读成 5
1. #433 tʰən 读成 5
1. #433 tʰən 读成 5
1. #448 la 读成 1
1. #449 la 读成 3
1. #458 lio 读成 4
1. #478 liau 读成 1
1. #492 liəu 读成 1
1. #564 ʦau 读成 3
1. #581 ʦõ 读成 4
1. #591 ʦan 音频缺失
1. #664 ʦʰoŋ 例字误用 PUA，应当为𢱤，同㨦
1. #673 si 读成 1
1. #679 sia 读成 6
1. #722 sian 读成 ʦian5
1. #725 sən 读成 ɕən
1. #920 ʨyn 瑕疵，不饱满
1. #963 ȵi 读成 2
1. #987 ɕi 例字中「依衣伊醫」都是 i1
1. #988 ɕi 例字中「椅已以矣」都是 i3
1. #989 ɕi 发音是 i2，例字除攜外也都是 i2
1. #990 ɕi 发音是 i4，例字除「戲繫」外也都是 i4
1. #992 ɕi 发音是 i6，例字除「隙泣」外也都是 i6
1. #998 ɕy 读成 4
1. #1005 ɕya 读成蝦 ɕia
1. #1078 kuə 读成 go6
1. #1221 xəu 读成类似 xoŋ
1. 所有的 oŋ（除了 moŋ/ŋoŋ）几乎都读成了 ən

## 参考

- [漢語多功能字庫 - 長沙音節表](http://humanum.arts.cuhk.edu.hk/Lexis/lexi-mf/dialectIndex.php?point=F)
- https://www.bilibili.com/video/BV1Ry4y1Y765
- [长沙方言音系研究综述](https://www.zhihu.com/xen/market/remix/paid_magazine/1396306359609081856)
