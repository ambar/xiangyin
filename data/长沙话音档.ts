import json from './raw/长沙话音档.修正.json'
import {
  CSToneNo,
  csToneNo2octetToneNo,
  csToneNo2toneValue,
  getToneLetter,
  ToneType,
  toneValue2csToneNo,
} from './tones'
import {NormResult, SchemaEntries} from './types'

type metaKey = 'disabled' | 'flawed' | 'comment' | 'corrected'
const schema = [
  ['號', (v: string | null) => v],
  ['声母', String],
  ['韵母', String],
  ['调类', String],
  ['调值', Number],
  ['例字', (v: string[]) => v],
  ['元', (v: Partial<Record<metaKey, string>>) => v],
] as const

type RawDataItem = SchemaEntries<typeof schema>
export type DataItem = RawDataItem & {
  长沙调序: number
}

const charGroup = new Map<string, DataItem[]>()

const setIfNotSet = (char: string, item: DataItem) => {
  if (char) {
    if (charGroup.has(char)) charGroup.get(char)!.push(item)
    else charGroup.set(char, [item])
  }
}

export const items = json.map((x) => {
  const item = Object.fromEntries(
    schema.map(([key, val], i) => [key, val(x[i])])
  ) as DataItem
  // 自定义增补
  item.长沙调序 = toneValue2csToneNo[item.调值]
  item.例字.forEach((c) => setIfNotSet(c, item))
  return item
})

export const query = (
  char: string,
  toneType: ToneType = 'CSToneNo'
): NormResult[] => {
  const items = charGroup.get(char)
  if (items) {
    return items.map((x) => {
      let t: string = x.调类
      let tone: string | number = t
      if (toneType === 'CSToneNo') {
        tone = x.长沙调序
      } else if (toneType === 'ToneLetter') {
        tone = getToneLetter(csToneNo2toneValue[x.长沙调序 as CSToneNo])
      } else if (toneType === 'ToneValue') {
        tone = csToneNo2toneValue[x.长沙调序 as CSToneNo]
      } else if (toneType === 'OctetToneNo') {
        tone = csToneNo2octetToneNo[x.长沙调序 as CSToneNo]
      } else if (toneType === 'ToneName') {
        // TODO: 统一简繁
        tone = t
      } else {
        tone = x.长沙调序
      }
      return {音: x.声母 + x.韵母, 调: tone, 释: ''}
    })
  }
  return []
}
