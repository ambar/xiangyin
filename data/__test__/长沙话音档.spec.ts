import {query} from '../长沙话音档'

test('query', () => {
  expect(query('中')).toMatchSnapshot()
  expect(query('椅')).toMatchSnapshot()
})
