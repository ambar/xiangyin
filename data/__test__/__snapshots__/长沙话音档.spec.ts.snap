// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query 1`] = `
Array [
  Object {
    "调": 1,
    "释": "",
    "音": "tʂoŋ",
  },
  Object {
    "调": 4,
    "释": "",
    "音": "tʂoŋ",
  },
]
`;

exports[`query 2`] = `
Array [
  Object {
    "调": 3,
    "释": "",
    "音": "i",
  },
]
`;
